File: android/src/main/kotlin/com/miguelruivo/flutter/plugin/countrycodes/country_codes_plus/CountryCodesPlugin.kt
package com.miguelruivo.flutter.plugin.countrycodes.country_codes_plus

import android.content.Context
import android.os.Build
import java.util.Locale
import androidx.annotation.NonNull;
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** CountryCodesPlugin */
public class CountryCodesPlugin: FlutterPlugin, MethodCallHandler {
  private lateinit var channel: MethodChannel
  private lateinit var applicationContext: Context

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    applicationContext = flutterPluginBinding.applicationContext
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "country_codes_plus")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {

    when (call.method) {
        "getLocale" -> {
          val locale = resolveLocale()
          result.success(listOf(locale.language, locale.country, getLocalizedCountryNames(call.arguments as String?)))
        }
        "getRegion" -> result.success(resolveLocale().country)
        "getLanguage" -> result.success(resolveLocale().language)
        else -> result.notImplemented()
    }
  }

  private fun getLocalizedCountryNames(localeTag: String?) : HashMap<String, String> {
    val localizedCountries: HashMap<String,String> = HashMap()
    val displayLocale = resolveDisplayLocale(localeTag)

    for (countryCode in Locale.getISOCountries()) {
      val locale = Locale("", countryCode)
      val countryName: String? = locale.getDisplayCountry(displayLocale)
      localizedCountries[countryCode.uppercase()] = countryName ?: ""
    }
    return localizedCountries
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  private fun resolveDisplayLocale(localeTag: String?): Locale {
    if (!localeTag.isNullOrBlank()) {
      return Locale.forLanguageTag(localeTag)
    }
    return resolveLocale()
  }

  private fun resolveLocale(): Locale {
    val config = applicationContext.resources.configuration
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
      val locales = config.locales
      for (i in 0 until locales.size()) {
        val locale = locales[i]
        if (locale != null && locale.country.isNotEmpty()) {
          return locale
        }
      }
      if (locales.size() > 0) {
        return locales[0]
      }
    } else {
      @Suppress("DEPRECATION")
      return config.locale ?: Locale.getDefault()
    }

    return Locale.getDefault()
  }
}

File: ios/country_codes_plus/Sources/country_codes_plus/CountryCodesPlugin.swift
import Flutter
import Foundation
import UIKit

public class CountryCodesPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "country_codes_plus", binaryMessenger: registrar.messenger())
    let instance = CountryCodesPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getRegion":
      result(Locale.autoupdatingCurrent.regionCode)
    case "getLanguage":
      result(Locale.autoupdatingCurrent.languageCode)
    case "getLocale":
      result([
        Locale.autoupdatingCurrent.languageCode as Any,
        Locale.autoupdatingCurrent.regionCode as Any,
        getLocalizedCountryNames(localeTag: call.arguments as? String),
      ])
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func getLocalizedCountryNames(localeTag: String?) -> [String: String] {
    var localizedCountries: [String: String] = [:]

    let displayLocale: Locale
    if let tag = localeTag, !tag.isEmpty {
      displayLocale = Locale(identifier: tag)
    } else {
      displayLocale = Locale.autoupdatingCurrent
    }

    for countryCode in NSLocale.isoCountryCodes {
      let countryName = (displayLocale as NSLocale).displayName(forKey: .countryCode, value: countryCode)
      localizedCountries[countryCode.uppercased()] = countryName ?? ""
    }
    return localizedCountries
  }
}
